src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+using EditorMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.SyntaxTree, Microsoft.CodeAnalysis.Editing.DocumentEditor>;
+using SyntaxMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxNode[]>;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal class MembersPuller
+    {
+        internal static readonly MembersPuller Instance = new MembersPuller();
+
+        private MembersPuller()
+        {
+        }
+
+        /// <summary>
+        /// Return the CodeAction to pull <paramref name="selectedMember"/> up to destinationType. If the pulling will cause error,
+        /// it will return null.
+        /// </summary>
+        public CodeAction TryComputeCodeAction(
+            Document document,
+            ISymbol selectedMember,
+            INamedTypeSymbol destination)
+        {
+            var result = PullMembersUpAnalysisBuilder.BuildAnalysisResult(destination, ImmutableArray.Create((member: selectedMember, makeAbstract: false)));
+            if (result.PullUpOperationCausesError ||
+                IsSelectedMemberDeclarationAlreadyInDestination(selectedMember, destination))
+            {
+                return null;
+            }
+
+            return new SolutionChangeAction(
+                string.Format(FeaturesResources.Add_to_0, result.Destination),
+                cancellationToken => PullMembersUpAsync(document, result, cancellationToken));
+        }
+
+        /// <summary>
+        /// Return the changed solution if all changes in result are applied.
+        /// </summary>
+        /// <param name="result">Contains the members to pull up and all the fix operations</param>>
+        internal async Task<Solution> PullMembersUpAsync(
+            Document document,
+            PullMembersUpAnalysisResult result,
+            CancellationToken cancellationToken)
+        {
+            if (result.Destination.TypeKind == TypeKind.Interface)
+            {
+                return await PullMembersIntoInterfaceAsync(document, result, cancellationToken).ConfigureAwait(false);
+            }
+            else if (result.Destination.TypeKind == TypeKind.Class)
+            {
+                return await PullMembersIntoClassAsync(document, result, cancellationToken).ConfigureAwait(false);
```

Comment:
this method does not need to be async, and we don't need await/configureawait in either of these cases.